tensorflow/core/kernels/split_v_op.cc

```diff
@@ -236,15 +287,32 @@ class SplitVOpCPUImpl {
             functor::Split<CPUDevice, T, NDims>()(
                 context->eigen_device<CPUDevice>(), result_shaped,
                 input_reshaped, current_indices, sizes);
+              }
+            }
           }
+        };
+    if (use_parallelism_between_outputs) {
+      const double ideal_load_per_thread =
+          input_reshaped.dimension(split_dim) / split_sizes_vec.size();
+      bool load_balancing = true;
+      for (auto split_size : split_sizes_vec) {
+        if (split_size / ideal_load_per_thread > 3) {
+          load_balancing = false;
         }
       }
-    };
-    if (use_parallelism_between_outputs) {
-      // Run in parallel, disabling parallelism in functor.
-      Shard(num_split,
-            context->device()->tensorflow_cpu_worker_threads()->workers,
-            num_split, input_element_count / num_split, range_output_func);
+      if ((!load_balancing && input_reshaped.dimension(0) > 32) && split_dim) {
+        // Each thread processes the same amount of data, and then copies data
+        // to all output tensors .
+        ParallelSplitByInputData(context, input_reshaped, input_shape,
```

Comment:
I don't think this actually belongs in `use_parallelism_between_outputs`, since that's not what it does.

Without `use_parallelism_between_outputs`, the split is done via sequential slices.  Each slice, however, internally *does* use multiple threads.  The key insight in this change seems to be that we don't need to wait for each slice to finish before starting the next slice.  I'm surprised it would make that much of a difference though, since each slice operation should make good use of the threadpool.

Or perhaps the issue is that we shouldn't actually be setting `use_parallelism_between_outputs` in cases where there is an imbalance in the split size.  Maybe only the condition where this is set above needs to be modified to account for the split imbalance, and in that case fall back to relying on sequential calls `functor::Split`.

Can you run some benchmarks to check if simply modifying the condition for `use_parallelism_between_outputs` based on your load-balancing heuristic would get the same speed improvements?  If so, that would simplify the implementation.  If not, we should probably replace the non-parallelism-between-outputs path with this modified approach.

Reply:
I run a simply benchmark, and I get:
original parallel：0.5327s
new parallel : 0.3560s
original serial slice: 0.75589s

My test code:
```
import tensorflow as tf
import time

count = 42
keys = tf.random.uniform(shape=[48000, 107, 8], dtype=tf.int64, maxval=int(1e8))
tmp = maxval=107 // count + 1
arr = []
for i in range(0,41):
   arr.append(2);
arr.append(25);
start = time.time()
for i in range(0,5):
  ret = tf.split(keys, arr, axis=1)

print('split_v seconds: %f' % (time.time() - start))
```

Reply:
It's same function within origin parallel and origin serial. 
Although it is possible to parallelize internally, only one result data is processed at a time.
In any case, one by one processing results, there will be unfriendly memory access problems.

The new method traverses the data sequentially and puts the data in multiple result sets.

Reply:
The original method does the following (in pseudocode):
- If `use_parallelism_between_outputs == true`,
  ```
  parallel_for i in range(0, num_outputs):
    output(i) = serial_split(input, indices(i), size(i))
  ```
- Otherwise,
  ```
  serial_for i in range(0, num_outputs):
    output(i) = parallel_split(input, indices(i), size(i))
  ```

I agree that there may be a better way to do this than sequential splits, and your approach seems logical.  Though I don't get the same speed-up you see.

I modified your benchmark slightly to use `timeit`:
```python
import tensorflow as tf
import timeit

keys = tf.random.uniform(shape=[48000, 107, 8], dtype=tf.int64, maxval=int(1e8))

arr = []
for i in range(0,41):
  arr.append(2)
arr.append(25)
  
def split_benchmark():
  ret = tf.split(keys, arr, axis=1)
  
timeit_iters = 10000
split_time = timeit.timeit(split_benchmark, number=timeit_iters)/timeit_iters
print('split_v seconds/iter: %f' % split_time)
```
Before your change: `split_v seconds/iter: 0.032396` 
After your change: `split_v seconds/iter: 0.034091`

I also tried running the c++ benchmarks in `split_v_op_test.cc`, but they segfault:
```
$ bazel run -c opt //tensorflow/core/kernels:split_v_op_test -- --benchmarks=all
...
*** BEGIN MANGLED STACK TRACE ***
/home/cantonios/.cache/bazel/_bazel_cantonios/4f63cee998cffde741ccb57af1c4cff3/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/core/kernels/split_v_op_test.runfiles/org_tensorflow/tensorflow/core/kernels/split_v_op_test(+0x216016)[0x55d6e79a5016]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x14140)[0x7fec26dcf140]
/lib/x86_64-linux-gnu/libc.so.6(+0x86ab9)[0x7fec26c4fab9]
/home/cantonios/.cache/bazel/_bazel_cantonios/4f63cee998cffde741ccb57af1c4cff3/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/core/kernels/../../../_solib_k8/_U_S_Stensorflow_Score_Skernels_Csplit_Uv_Uop_Utest___Utensorflow/libtensorflow_framework.so.2(+0x9bdea1)[0x7fec2be42ea1]
/home/cantonios/.cache/bazel/_bazel_cantonios/4f63cee998cffde741ccb57af1c4cff3/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/core/kernels/../../../_solib_k8/_U_S_Stensorflow_Score_Skernels_Csplit_Uv_Uop_Utest___Utensorflow/libtensorflow_framework.so.2(+0x9c4262)[0x7fec2be49262]
/home/cantonios/.cache/bazel/_bazel_cantonios/4f63cee998cffde741ccb57af1c4cff3/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/core/kernels/../../../_solib_k8/_U_S_Stensorflow_Score_Skernels_Csplit_Uv_Uop_Utest___Utensorflow/libtensorflow_framework.so.2(_ZN5Eigen15ThreadPoolTemplIN10tensorflow6thread16EigenEnvironmentEE10WorkerLoopEi+0x1b9)[0x7fec2c2e70c9]
/home/cantonios/.cache/bazel/_bazel_cantonios/4f63cee998cffde741ccb57af1c4cff3/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/core/kernels/../../../_solib_k8/_U_S_Stensorflow_Score_Skernels_Csplit_Uv_Uop_Utest___Utensorflow/libtensorflow_framework.so.2(_ZNSt17_Function_handlerIFvvEZN10tensorflow6thread16EigenEnvironment12CreateThreadESt8functionIS0_EEUlvE_E9_M_invokeERKSt9_Any_data+0x48)[0x7fec2c2e3d88]
/home/cantonios/.cache/bazel/_bazel_cantonios/4f63cee998cffde741ccb57af1c4cff3/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/core/kernels/../../../_solib_k8/_U_S_Stensorflow_Score_Skernels_Csplit_Uv_Uop_Utest___Utensorflow/libtensorflow_framework.so.2(+0xe451e0)[0x7fec2c2ca1e0]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x8ea7)[0x7fec26dc3ea7]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f)[0x7fec26cc6def]
*** END MANGLED STACK TRACE ***
```
These benchmarks run successfully without your change.

Reply:
The above example just proves that the time consumption of origin parallel, origin serial and new parallel are different.
I'm sorry that  I used the method of forcibly changing the code to verify the difference between the above three methods.
```before your change and after your change is same``` If the code is not changed forcibly, the example will run serially.

If you want to see the difference between changes and no changes, please work hard to run the following example, thx
```
import tensorflow as tf
import timeit

keys = tf.random.uniform(shape=[6000, 107, 8], dtype=tf.int64, maxval=int(1e8))

arr = []
for i in range(0,41):
  arr.append(2)
arr.append(25)

def split_benchmark():
  ret = tf.split(keys, arr, axis=1)

timeit_iters = 10000
split_time = timeit.timeit(split_benchmark, number=timeit_iters)/timeit_iters
print('split_v seconds/iter: %f' % split_time)
```
my result:
new parallel function : split_v seconds/iter: ```0.007545```
origin parallel function : split_v seconds/iter: ```0.012376```
```

Reply:
I'm still getting segfaults with 
```
bazel run -c opt //tensorflow/core/kernels:split_v_op_test -- --benchmarks=all
```
---
And to re-iterate my first comment in this thread: I don't think you want to do this for the case `use_parallelism_between_outputs == true` (or at least not ***only*** for that case).  There's a limit for that s.t. if the tensor is too big, it reverts back to the serial split approach - and I think that's were your method actually performs better.

For example, if I hack the code to always use your approach, then I get:
| input shape | `[6000, 107, 8]`  | `[48000, 107, 8]` | `[196000, 107, 8]` |
|-------|------------------|-------------------|-------------------|
|orig:    | 0.004986 | 0.036530 | 0.144675 |
|new + hack:   | 0.002297  | 0.013596 | 0.054090 |

We probably want to do your suggested split by default, except in some specific cases where the current `use_parallelism_between_outputs` wins out.

Reply:
I changed the select branch code.
```
// Default function. The new method performs better when there are more data on the 0th dimension.
if (large dataset && dim0_size > 32 && split_dim != 0) {
    new parallel split function .
}
// The origin parallel split method is more suitable for when there are more output results.
else if (use_parallelism_between_outputs) {
    origin parallel split function
}
// If the first two bad cases are hit, the sequential splitting method is used
else {
    origin serialize split function
}
```

```Note that for the first case, which is the only one the new approach is actually used without the hack, I don't see any improvement.``` I'm sorry. I fixed a bug, please retry.